src/chunkserver/bgjobs.cc

```diff
@@ -41,609 +29,360 @@
+	statusQueue = std::make_unique<ProducerConsumerQueue>();
 
-	if (jp->statusQueue->isEmpty()) {   // first status
-		eassert(write(jp->wpipe,&status,1)==1); // write anything to wake up select
+	for (uint8_t i = 0; i < workers; ++i) {
+		workerThreads.emplace_back(&JobPool::workerThread, this);
 	}
-	jp->statusQueue->put(jobid, status, nullptr, 1);
 }
 
-static inline int job_receive_status(jobpool *jp,uint32_t *jobid,uint8_t *status) {
+JobPool::~JobPool() {
+	for (uint8_t i = 0; i < workers; ++i) {
+		jobsQueue->put(0, JobPool::ChunkOperation::Exit, nullptr, 1);
+	}
+
+	for (auto &thread : workerThreads) {
+		if (thread.joinable()) { thread.join(); }
+	}
+
+	if (!statusQueue->isEmpty()) { checkJobs(); }
+
+	jobsQueue.reset();
+	statusQueue.reset();
+
+	workerThreads.clear();
+
+	close(rpipe);
+	close(wpipe);
+}
+
+uint32_t JobPool::addJob(ChunkOperation operation, JobCallback callback, void *extra,
+                         ProcessJobCallback processJob) {
+	std::unique_lock lock(jobsMutex);
+	uint32_t jobId = nextJobId++;
+	auto job = std::make_unique<Job>();
+	job->jobId = jobId;
+	job->callback = std::move(callback);
+	job->processJob = std::move(processJob);
+	job->extra = extra;
+	job->state = JobPool::State::Enabled;
+	jobHash[jobId] = std::move(job);
+	jobsQueue->put(jobId, operation, reinterpret_cast<uint8_t *>(jobHash[jobId].get()), 1);
+	return jobId;
+}
+
+uint32_t JobPool::getJobCount() const {
 	TRACETHIS();
-	uint32_t qstatus;
+	return jobsQueue->elements();
+}
 
-	std::lock_guard pipeLockGuard(jp->pipeMutex);
+void JobPool::disableAndChangeCallbackAll(const JobCallback& callback) {
+	std::lock_guard jobsLockGuard(jobsMutex);
+	for (auto &[jobId, job] : jobHash) {
+		if (job->state == JobPool::State::Enabled) { job->state = JobPool::State::Disabled; }
+		job->callback = callback;
+	}
+}
 
-	jp->statusQueue->get(jobid, &qstatus, nullptr, nullptr);
-	*status = qstatus;
-	PRINTTHIS(*jobid);
-	PRINTTHIS((int)*status);
-	if (jp->statusQueue->isEmpty()) {
-		eassert(read(jp->rpipe,&qstatus,1)==1); // make pipe empty
-		return 0;       // last element
+void JobPool::disableJob(uint32_t jobId) {
+	std::unique_lock jobsUniqueLock(jobsMutex, std::defer_lock);
+	auto jobIterator = jobHash.find(jobId);
+	if (jobIterator != jobHash.end()) {
+		jobsUniqueLock.lock();
+		if (jobIterator->second->state == JobPool::State::Enabled) {
+			jobIterator->second->state = JobPool::State::Disabled;
+		}
+		jobsUniqueLock.unlock();
 	}
-	return 1;       // not last
 }
 
-void* job_worker(void *th_arg) {
-	TRACETHIS();
+void JobPool::checkJobs() {
+	uint32_t jobId;
+	uint8_t status;
+	int notLastJob = 1;
+	do {
+		notLastJob = receiveStatus(jobId, status);
+		auto jobIterator = jobHash.find(jobId);
+		if (jobIterator != jobHash.end()) {
+			auto callback = jobIterator->second->callback;
+			if (callback) { callback(status, jobIterator->second->extra); }
+			jobHash.erase(jobIterator);
+		}
+	} while (notLastJob > 0);
```

Comment:
Consider changing this by:
```suggestion
	} while (notLastJob);
```
this is just an optional suggestion in case it will fit future behavior of `receiveStatus` function.

Reply:
This was my first implementation. I changed to avoid clangd warning `Implicit conversion 'int' -> 'bool'`.

Reply:
Then, why not change the type as well, altogether with the return type of `receiveStatus`?

Reply:
I changed to `bool` the type of `notLastJob` as well as the return type of `receiveStatus`, as both suggested.